src/modules/be-water-smart/be-water-smart.component.ts

```diff
@@ -0,0 +1,686 @@
+        selectedMeter,
+        this.createSubMeterList(selectedMeters),
+      )
+      .subscribe({
+        next: response => {
+          if (response.hasOwnProperty("virtualMeterId")) {
+            this.extractVMeters();
+            this.selectedPhysicalMeters = [];
+            this.selectedVirtualMeters = [];
+            this.newVMeterName = "";
+            this.newSuperMeterName = "";
+          }
+        },
+        error: error => {
+          console.log(error);
+        },
+      });
+  }
+
+  /**
+   * help function for addVMeter()
+   * @returns a list of all ids which are inside the virtual meter
+   */
+  createSubMeterList(selectedMeters: any): object {
+    let id_list: string[] = [];
+
+    selectedMeters.forEach((item: {id: string}) => {
+      id_list.push(item.id);
+    });
+
+    return {submeterIds: id_list};
+  }
+
+  /**
+   * delete a virtual meter
+   * @param id  name of virtual meter, which functions as it's id
+   * @param index index of meter in arr, to hotreload page
+   */
+  deleteVMeterById(id: string, index: number): void {
+    if (this.isDeleting) {
+      return;
+    }
+
+    // flag true aslong as deletion is processed)
+    this.isDeleting = true;
+
+    let tmp = this.vMeters.splice(index, 1);
+
+    console.log(tmp);
+
+    this.bwsService.delVirtualMeterById(id).subscribe({
+      next: response => {
+        if (response && response.hasOwnProperty("msg")) {
+          this.vMeters.push(tmp[0]);
+          alert(
+            "Virtual Meter with Name " +
+              id +
+              " not found and can not be deleted!",
+          );
+        }
+      },
+      error: error => {
+        console.log(error);
+      },
+      complete: () => {
+        this.isDeleting = false;
+      },
+    });
+  }
+
+  // ---------- Algorithm Functions ----------
+
+  /**
+   * train one of the Models and retrieve the training data
+   */
+  trainModel(): void {
+    if (this.choiceVirtualMeter) {
+      this.selectedVirtualMeter = this.getVirtualMeter(this.choiceVirtualMeter);
+    }
+
+    if (!this.selectedVirtualMeter) {
+      console.log("No Virtual Meter detected!");
+      return;
+    }
+
+    if (this.choiceAlgorithm) {
+      this.selectedAlgorithm = this.getAlgorithm(this.choiceAlgorithm);
+    }
+
+    if (!this.selectedAlgorithm) {
+      console.log("No algorithm detected!");
+      return;
+    }
+
+    if (!this.modelComment) {
+      alert("a comment is necessary!");
+      return;
+    }
+
+    this.bwsService
+      .putTrainModel(
+        this.selectedVirtualMeter,
+        this.selectedAlgorithm,
+        this.modelComment,
+      )
+      .subscribe({
+        next: () => {
+          this.extractModels();
+          this.choiceAlgorithm = undefined;
+          this.selectedAlgorithm = undefined;
+          this.selectedVirtualMeter = undefined;
+          this.modelComment = undefined;
+        },
+        error: error => {
+          console.log(error);
+        },
+      });
```

Comment:
you don't need to use `next` and `error` if you just want to log errors, or is just logging the error the desired behavior?